ios/Classes/HLSCachingReverseProxyServer.swift

```diff
@@ -0,0 +1,198 @@
+import GCDWebServer
+import PINCache
+
+open class HLSCachingReverseProxyServer {
+  static let originURLKey = "__hls_origin_url"
+
+  private let webServer: GCDWebServer
+  private let urlSession: URLSession
+  private let cache: PINCaching
+
+  private(set) var port: Int?
+
+  public init(webServer: GCDWebServer, urlSession: URLSession, cache: PINCaching) {
+    self.webServer = webServer
+    self.urlSession = urlSession
+    self.cache = cache
+
+    self.addRequestHandlers()
+  }
+
+  // MARK: Starting and Stopping Server
+
+  open func start(port: UInt) {
+    guard !self.webServer.isRunning else { return }
+    self.port = Int(port)
+    self.webServer.start(withPort: port, bonjourName: nil)
+  }
+
+  open func stop() {
+    guard self.webServer.isRunning else { return }
+    self.port = nil
+    self.webServer.stop()
+  }
+
+
+  // MARK: Resource URL
+
+  open func reverseProxyURL(from originURL: URL) -> URL? {
+    guard let port = self.port else { return nil }
+
+    guard var components = URLComponents(url: originURL, resolvingAgainstBaseURL: false) else { return nil }
+    components.scheme = "http"
+    components.host = "127.0.0.1"
+    components.port = port
+
+    let originURLQueryItem = URLQueryItem(name: Self.originURLKey, value: originURL.absoluteString)
+    components.queryItems = (components.queryItems ?? []) + [originURLQueryItem]
+
+    return components.url
+  }
+
+
+  // MARK: Request Handler
+
+  private func addRequestHandlers() {
+      print("\(Date()) rpc: Adding request handlers")
+    self.addPlaylistHandler()
+    self.addSegmentHandler()
+  }
+
+  private func addPlaylistHandler() {
+    self.webServer.addHandler(forMethod: "GET", pathRegex: "^/.*\\.m3u8$", request: GCDWebServerRequest.self) { [weak self] request, completion in
+      print("\(Date()) rpc: Received request for playlist: \(request.url.path)")
+      
+      guard let self = self else {
+        return completion(GCDWebServerErrorResponse(statusCode: 410))
```

Comment:
I don't know the syntax for constants in Swift, but it would be good to do something named to give a little more context as to why we always return a 410.
```
const FORCED_FALLBACK_STATUS_CODE = 410;
```